[2661] Fix BroadcastReceiver ANR at ScreenOffTrigger

FranAguilera opened the pull request:
Resolves https://github.com/square/leakcanary/issues/2661

**Issue**

BroadcastReceiver#onReceive runs by default on the main thread. If onReceive doesn't complete in a timely manner, the system will report an Application Not Responding 

We've seen instances of this occurring frequently on internal builds

**Solution**

NOTE: Kudos to @pyricau for the guidance on adding a delay. 

Let the broadcast receiver complete immediately and schedule the executor with an initial delay to avoid issues with the broadcast receiver not completing

**Verification**

Verified that the the ANR no longer occurs and reporting still completes without issues on one of our main apps (internal release build)

This was heavily tested on internal builds

pyricau commented on leakcanary/leakcanary-android-release/src/main/java/leakcanary/ScreenOffTrigger.kt:
```diff
@@ -42,20 +42,20 @@ class ScreenOffTrigger(
-            analysisClient.newJob(JobContext(ScreenOffTrigger::class))
-          currentJob = job
-          analysisExecutor.execute {
+      analysisExecutor.execute {
```
If `analysisExecutor.execute` runs the task quickly enough, there's a chance that we might dump the heap & freeze the VM before the process has had a chance to report back to `system_server`, right?

Looks to me like if we want to avoid that, we need to execute with a delay instead"?

Jlb252423 replied:
How is this complaint or effective in this work space or the relevance in what I do

FranAguilera committed:
Fix BroadcastReceiver ANR

FranAguilera committed:
PR feedback

Adding delay

Add a delayedScheduledExecutorService

Fix BroadcastReceiver ANR

PR feedback

Fix typo

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/square/leakcanary?pullRequest=2732) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/square/leakcanary?pullRequest=2732) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/square/leakcanary?pullRequest=2732) it.</sub>

pyricau commented on leakcanary/leakcanary-android-release/src/main/java/leakcanary/ScreenOffTrigger.kt:
```diff
@@ -78,4 +91,38 @@ class ScreenOffTrigger(
+
+  private class DelayedScheduledExecutorService(
+    private val analysisExecutorDelayMillis:Long,
+    private val analysisExecutor: Executor){
```
lint will probs get mad at you here, format the file.

pyricau commented on leakcanary/leakcanary-android-release/src/main/java/leakcanary/ScreenOffTrigger.kt:
```diff
@@ -32,30 +37,38 @@ class ScreenOffTrigger(
-  private var currentJob: HeapAnalysisJob? = null
+  private val currentJob = AtomicReference<HeapAnalysisJob?>()
 
+  private val delayedScheduledExecutorService = DelayedScheduledExecutorService()
```
I only see a constructor with 2 params in the definition, but no params here?

FranAguilera replied:
uops my bad, I missed to port this

pyricau commented on leakcanary/leakcanary-android-release/src/main/java/leakcanary/ScreenOffTrigger.kt:
```diff
@@ -78,4 +91,38 @@ class ScreenOffTrigger(
+  }
+
+  companion object{
+    const val INITIAL_EXECUTOR_DELAY_IN_MILLI = 500L
```
nit: `_IN_MILLI` => `MS` or `MILLIS`

pyricau commented on leakcanary/leakcanary-android-release/src/main/java/leakcanary/ScreenOffTrigger.kt:
```diff
@@ -32,30 +37,38 @@ class ScreenOffTrigger(
-            currentJob = null
-            analysisCallback(result)
+          if(currentJob.compareAndSet(null, job)){
+            delayedScheduledExecutorService.schedule {
```
Couldn't this have been a simple `Handler(Looper.getMainLooper()).postDelayed({}, 500)` ?

pyricau replied:
(you can cancel runnables on handler just fine)

FranAguilera replied:
Good point, simplified to use Handler

Jlb252423 replied:
I'm not being handled by no one the fact that you use this term is concerning

pyricau commented on leakcanary/leakcanary-android-release/src/main/java/leakcanary/ScreenOffTrigger.kt:
```diff
@@ -19,7 +24,7 @@ class ScreenOffTrigger(
    * This should likely be a single thread executor with a background thread priority.
    */
-  private val analysisExecutor: Executor,
+  analysisExecutor: Executor,
```
why not private? seems wrong?

FranAguilera replied:
My bad, this was a mistake from previously porting changes

pyricau commented on leakcanary/leakcanary-android-release/src/main/java/leakcanary/ScreenOffTrigger.kt:
```diff
@@ -32,30 +35,39 @@ class ScreenOffTrigger(
+   *
+   * If not specified, the initial delay is 500 ms
+   */
+  private val analysisExecutorDelayMillis: Long = INITIAL_EXECUTOR_DELAY_MILLIS
```
nit: take in a `Duration` instead.

FranAguilera replied:
good callout, updated now

pyricau commented on leakcanary/leakcanary-android-release/src/main/java/leakcanary/ScreenOffTrigger.kt:
```diff
@@ -32,30 +35,39 @@ class ScreenOffTrigger(
-  private var currentJob: HeapAnalysisJob? = null
+  private val currentJob = AtomicReference<HeapAnalysisJob?>()
+  private val analysisHandler = Handler(Looper.getMainLooper())
+  private var analysisRunnable: Runnable? = null
```
nit: rename this to `triggerAnalysis`?

pyricau commented:
Context from the past: the key issue here wasn't that we were actively blocking the main thread, but rather that we were starting the analysis immediately, which triggered a heap dump, so the VM froze and the receiver wouldn't respond in time.
